### ISSUE TITLE
Unexpected error when narrowing TypeVar

### ISSUE THREAD

**Issue by pganssle** [python/mypy#14045]
# Bug report

I ran into this recently and @glyph came up with [a very elegant minimal working example](https://mypy-play.net/?mypy=latest&python=3.10&gist=20456ce8ac2d90b3b3df82cc5a93a384):

```python
from typing import TypeVar

T = TypeVar("T")

def f(s: T) -> T:
    if isinstance(s, str):
        return s # main.py:7: error: Incompatible return value type (got "str", expected "T")
    return s
```

Basically, if you have a function that takes a typevar and you narrow it using an `isinstance` check, `mypy` decides that the concrete type does not match the generic type from the signature, despite the fact that the concrete type is just the narrowed generic.

The best workaround I've found is to return `cast(T, s)`. Glyph and I both tried playing around with overloads and couldn't find something to work around the issue other than just a cast.

**Comment by glyph** [comment]
I should note that there are some other cases which make this complicated:

```python3
from typing import TypeVar, reveal_type

T = TypeVar("T")

def f(a: T, b: T) -> T:
    return a

reveal_type(f("1", 2))
```

`T` is `object` in this call, due to that being the common supertype of the two arguments.  So the TypeVar itself, T, is actually *widened*, to include `str` but not necessarily to exclude anything else.  For example, if `isinstance(a, str)` is true, we can say nothing about `b`; but we *can* say that another `str` should be an acceptable return type.

Ooh. Except, maybe not:

```python3
from typing import TypeVar, reveal_type, NewType

T = TypeVar("T")

Special = NewType("Special", str)
TheSpecial = Special("Hello")

def f(a: T) -> T:
    if isinstance(a, str):
        return "Goodbye"
    return a

reveal_type(f(TheSpecial))
```

If `"Hello"` is the only `Special` string allowed, then the runtime check tells us nothing here.

Maybe this is not trivially possible.

**Comment by sterliakov** [comment]
Is it the same issue as #13956?

**Comment by glyph** [comment]
> Is it the same issue as #13956?

I believe so.

**Comment by hauntsaninja** [comment]
Closing as duplicate of #10817
